build(cmake): simplify instructions (`cmake -B build && cmake --build build ...`)

ochafik opened the pull request:
- Updated instructions & CI to skip the `mkdir -p build && cd build` steps using config-time flag `-B` and build-time flag `--build` 

  ```bash
  mkdir -b build
  cd build
  cmake ..
  make # or `cmake --build . --config Release`
  ```

  Becomes:

  ```bash
  cmake -B build
  cmake --build build --config Release
  # --config flag is ignored for Makefile generation (default) but crucial for Release builds w/ Xcode / MSVC, as pointed out by reviewers
  ```

- Documented how to do `Debug` builds

- ~~Removed `--config Release` from most cmake build commands (except for Visual Studio) as `Release` is the [default](https://github.com/ggerganov/llama.cpp/blob/7bb36ccf91b8a2e92b182dd75624f1fd7cb205ac/CMakeLists.txt#L8) and the `--config` flag is [often useless](https://stackoverflow.com/questions/7724569/debug-vs-release-in-cmake/64719718#64719718) anyway)~~

  **Note**: To build in Debug w/ CMake one generally needs to set `-DCMAKE_BUILD_TYPE=Debug` in the "config" command (`--config Debug` on the build command works *only* for [Ninja Multi-Config](https://cmake.org/cmake/help/latest/generator/Ninja%20Multi-Config.html), XCode or Visual Studio generators)

Commit:
readme: cmake . -B build && cmake --build build

cebtenzzre commented on README.md:
```diff
@@ -486,10 +480,8 @@ Building the program with BLAS support may lead to some performance improvements
-    cmake .. -DLLAMA_CUDA=ON
-    cmake --build . --config Release
+    cmake . -B build -DLLAMA_CUDA=ON
+    cmake --build build .
```
```suggestion
    cmake --build build
```

cebtenzzre commented:
The source directory defaults to the current directory when using `-B`. So `cmake . -B build` should just be `cmake -B build`.

ochafik committed:
build: fix typo

Co-authored-by: Jared Van Bortel <cebtenzzre@gmail.com>

Commit:
build: drop implicit . from cmake config command

Commit:
build: remove another superfluous .

Commit:
build: update MinGW cmake commands

ochafik commented:
> The source directory defaults to the current directory when using -B. So cmake . -B build should just be cmake -B build.

Neat! Updated.

NeoZhangJianyu commented on README-sycl.md:
```diff
@@ -227,16 +226,15 @@ Similarly, user targeting Nvidia GPUs should expect at least one SYCL-CUDA devic
-#build all binary
-cmake --build . --config Release -j -v
+# build all binary
+cmake --build build -j -v
```
```suggestion
cmake --build . --config Release -j -v
```
Thank you to enhance the build. 
I test it to work well for SYCL backend on Intel GPU.
I suggest keeping the old format, which is explicit to show how to build Release. And user know how to build with Debug mode easily.
This is guide. We should show easy and more info to user as possible.

ochafik replied:
Thanks for testing this out (and for catching the other typo!!)

The old format is actually misleading as ‘—config Debug’ would also build in Release mode (only builds Debug for multi config generators such as Visual Studio, XCode or Ninja Multi-Config, not for Makefile or normal Ninja modes). To build in debug *in general*, you need to set ‘-DCMAKE_BUILD_TYPE=Debug’ on the first “config” cmake command. I’ve hinted about this in the main readme but maybe I should expand a bit / add this to each of the readmes?

ochafik replied:
Reverted the `--config Release` flag (as it's not the default for MSVC / Xcode etc) and documented how to do Debug builds on the main Readme.

NeoZhangJianyu commented on README-sycl.md:
```diff
@@ -412,17 +409,15 @@ b. Download & install mingw-w64 make for Windows provided by w64devkit
 
 # Option 1: Use FP32 (recommended for better performance in most cases)
-cmake -G "MinGW Makefiles" ..  -DLLAMA_SYCL=ON -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icx  -DCMAKE_BUILD_TYPE=Release
+cmake -B build -G "MinGW Makefiles" -DLLAMA_SYCL=ON -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icx  -DCMAKE_BUILD_TYPE=Release
```
This line is fault in windows build.
Please fix it or keep old format.
```
CMake Warning:
  Ignoring extra path from command line:

   "MinGW Makefiles"


CMake Error: The source directory "C:/Users/jianyuzh/OneDrive - Intel Corporation/Documents/zjy/demo/llama.cpp/arthw/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
comomand error: 1
```

NeoZhangJianyu commented on README-sycl.md:
```diff
@@ -412,17 +409,15 @@ b. Download & install mingw-w64 make for Windows provided by w64devkit
 
 # Option 2: Or FP16
-cmake -G "MinGW Makefiles" ..  -DLLAMA_SYCL=ON -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icx  -DCMAKE_BUILD_TYPE=Release -DLLAMA_SYCL_F16=ON
+cmake -B build "MinGW Makefiles" -DLLAMA_SYCL=ON -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icx  -DCMAKE_BUILD_TYPE=Release -DLLAMA_SYCL_F16=ON
```
This line is fault in windows build.
Please fix it or keep old format.

ochafik committed:
Update README-sycl.md

Co-authored-by: Neo Zhang Jianyu <jianyu.zhang@intel.com>

slaren commented:
> * Removed `--config Release` from most cmake build commands (except for Visual Studio) as `Release` is the [default](https://github.com/ggerganov/llama.cpp/blob/7bb36ccf91b8a2e92b182dd75624f1fd7cb205ac/CMakeLists.txt#L8) and the `--config` flag is [often useless](https://stackoverflow.com/questions/7724569/debug-vs-release-in-cmake/64719718#64719718) anyway)

I am not sure that this is entirely correct. `CMAKE_BUILD_TYPE` configures the build type for single-config generators, while `--config` selects the build type for multi-config generators (ie. MSVC). I have been trying to find what is the default value of `--config`, but I have not been able to find any references. Are you sure that removing `--config` will still build in release with MSVC in all cases?

Commit:
build: reinstate --config Release as not the default w/ some generators + document how to build Debug

Commit:
build: revert more --config Release

ochafik commented:
> > Removed --config Release from most cmake build commands (except for Visual Studio) as Release is the default and the --config flag is often useless anyway)

> I am not sure that this is entirely correct

@slaren Oops I misspoke, it's the default for single-generator configs (as set [here](https://github.com/ggerganov/llama.cpp/blob/ffe666572f98a686b17a2cd1dbf4c0a982e5ac0a/CMakeLists.txt#L8)), and the default is generator dependent for the others.

> I just ran a quick test, and without --config Release, it will result in a debug build. Therefore I think it is important to keep the --config Release, at least in the documentation that is meant to be applicable for all the platforms.

Thanks a lot for catching this! I just checked if setting `-DCMAKE_CONFIGURATION_TYPES=Release;Debug;MinSizeRel;RelWithDebInfo` would give control over the default (hoping it'd pick the first) but it's not the case at least for Xcode.

I've reverted the `--config Release` bit and added instructions for `Debug` builds on the main README, PTAL 😁

slaren commented on README.md:
````diff
@@ -321,12 +323,26 @@ In order to build llama.cpp you have three different options.
 
+    **Note**: for `Debug` builds, there are two cases:
+
+    - General case (esp. for default Makefile or Ninja generation):
+
+      ```bash
+      cmake -B build -DCMAKE_BUILD_TYPE=Debug
+      cmake --build build
+      ```
+
+    - Special case for multi-config generators (`-G` param set to Visual Studio, XCode...; note that `--config` is ignored by other generators):
````
Is there any reference that considers single-config generators as the general case, and multi-config generators as a special case? Otherwise, I would prefer more neutral language (ie. one case for single-config generators, and another case for multi-config generators, without additional qualifiers).

ochafik replied:
Good point, reworded this, thanks!

slaren commented on README.md:
````diff
@@ -517,7 +527,7 @@ Building the program with BLAS support may lead to some performance improvements
     ```bash
     CC=/opt/rocm/llvm/bin/clang CXX=/opt/rocm/llvm/bin/clang++ \
         cmake -H. -Bbuild -DLLAMA_HIPBLAS=ON -DAMDGPU_TARGETS=gfx1030 -DCMAKE_BUILD_TYPE=Release \
-        && cmake --build build -- -j 16
+        && cmake --build build --config Release -- -j 16
````
Not directly related to this change, but do you know what the `-H.` is meant to do here? According to the man page, `-H` is "Print usage information and exit".

ochafik replied:
Ah, looks like `cmake -H` will do that but `-H` is ignored when there are other args... and `-H.` is treated as `-H .` == `-H -S .`: I've now removed them all.

Commit:
build: nit / remove -H from cmake example

Commit:
build: reword debug instructions around single/multi config split
